### PR TITLE
[ENHANCEMENT] add record related webhook events

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

This PR adds record-related webhook events: `created`, `updated`, `deleted` and `completed`

**Type of change**

- Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: argilla-server/src/argilla_server/api/webhooks/v1/enums.py [hunk 1]
```diff
@@ -48,3 +53,13 @@ class ResponseEvent(str, Enum):
 
     def __str__(self):
         return str(self.value)
+
+
+class RecordEvent(str, Enum):
+    created = WebhookEvent.record_created.value
+    updated = WebhookEvent.record_updated.value
+    deleted = WebhookEvent.record_deleted
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/webhooks/v1/enums.py]:
```suggestion
    deleted = WebhookEvent.record_deleted.value
```

File: argilla-server/src/argilla_server/api/webhooks/v1/records.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+from argilla_server.api.webhooks.v1.schemas import RecordEventSchema
+from argilla_server.jobs.webhook_jobs import enqueue_notify_events
+from argilla_server.models import Record, Dataset
+
+
+async def notify_record_event(db: AsyncSession, record_event: RecordEvent, record: Record) -> List[Job]:
+    if record_event == RecordEvent.deleted:
+        return await _notify_record_deleted_event(db, record)
+
+    await Record.get(
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/webhooks/v1/records.py]: I'm not using `get` classmethod in the other functions. I'm simply doing a `db.execute`.

For example for datasets I'm doing the following:

```python
    # NOTE: Force loading required association resources required by the event schema
    (
        await db.execute(
            select(Dataset)
            .where(Dataset.id == dataset.id)
            .options(
                selectinload(Dataset.workspace),
                selectinload(Dataset.questions),
                selectinload(Dataset.fields),
                selectinload(Dataset.metadata_properties),
                selectinload(Dataset.vectors_settings),
            )
        )
    ).scalar_one()
```

File: argilla-server/src/argilla_server/api/webhooks/v1/records.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    )
+
+    event = RecordEventSchema.from_orm(record)
+
+    return await enqueue_notify_events(
+        db,
+        event=record_event,
+        timestamp=datetime.utcnow(),
+        data=jsonable_encoder(event),
+    )
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/webhooks/v1/records.py]:
```suggestion
    return await enqueue_notify_events(
        db,
        event=record_event,
        timestamp=datetime.utcnow(),
        data=RecordEventSchema.from_orm(record).dict(),
    )
```

File: argilla-server/src/argilla_server/api/webhooks/v1/records.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from datetime import datetime
+from typing import List
+
+from fastapi.encoders import jsonable_encoder
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/webhooks/v1/records.py]:
```suggestion
```

File: argilla-server/src/argilla_server/bulk/records_bulk.py [hunk 5]
```diff
@@ -29,6 +29,8 @@
 )
 from argilla_server.api.schemas.v1.responses import UserResponseCreate
 from argilla_server.api.schemas.v1.suggestions import SuggestionCreate
+from argilla_server.api.webhooks.v1.enums import RecordEvent
+from argilla_server.api.webhooks.v1.records import notify_record_event
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/bulk/records_bulk.py]: I would add an alias to explicitly see that it's version 1:

```suggestion
from argilla_server.api.webhooks.v1.records import notify_record_event as notify_record_event_v1
```

File: argilla-server/src/argilla_server/bulk/records_bulk.py [hunk 6]
```diff
@@ -73,6 +75,9 @@ async def create_records_bulk(self, dataset: Dataset, bulk_create: RecordsBulkCr
 
         await self._db.commit()
 
+        for record in records:
+            await notify_record_event(self._db, RecordEvent.created, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/bulk/records_bulk.py]:
```suggestion
            await notify_record_event_v1(self._db, RecordEvent.created, record)
```

File: argilla-server/src/argilla_server/bulk/records_bulk.py [hunk 7]
```diff
@@ -233,6 +240,13 @@ async def _fetch_existing_dataset_records(
 
         return {**records_by_external_id, **records_by_id}
 
+    async def _notify_record_events(self, records: List[Record]) -> None:
+        for record in records:
+            if record.inserted_at == record.updated_at:
+                await notify_record_event(self._db, RecordEvent.created, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/bulk/records_bulk.py]:
```suggestion
                await notify_record_event_v1(self._db, RecordEvent.created, record)
```

File: argilla-server/src/argilla_server/bulk/records_bulk.py [hunk 8]
```diff
@@ -233,6 +240,13 @@ async def _fetch_existing_dataset_records(
 
         return {**records_by_external_id, **records_by_id}
 
+    async def _notify_record_events(self, records: List[Record]) -> None:
+        for record in records:
+            if record.inserted_at == record.updated_at:
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/bulk/records_bulk.py:245]: 😮

File: argilla-server/src/argilla_server/bulk/records_bulk.py [hunk 9]
```diff
@@ -233,6 +240,13 @@ async def _fetch_existing_dataset_records(
...
+                await notify_record_event(self._db, RecordEvent.created, record)
+            else:
+                await notify_record_event(self._db, RecordEvent.updated, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/bulk/records_bulk.py]:
```suggestion
                await notify_record_event_v1(self._db, RecordEvent.updated, record)
```

File: argilla-server/src/argilla_server/contexts/datasets.py [hunk 10]
```diff
@@ -60,7 +60,8 @@
     VectorSettingsCreate,
 )
 from argilla_server.api.schemas.v1.vectors import Vector as VectorSchema
-from argilla_server.api.webhooks.v1.enums import DatasetEvent, ResponseEvent
+from argilla_server.api.webhooks.v1.enums import DatasetEvent, ResponseEvent, RecordEvent
+from argilla_server.api.webhooks.v1.records import notify_record_event
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/datasets.py]:
```suggestion
from argilla_server.api.webhooks.v1.records import notify_record_event as notify_record_event_v1
```

File: argilla-server/src/argilla_server/contexts/datasets.py [hunk 11]
```diff
@@ -837,6 +838,8 @@ async def update_record(
 
     await db.commit()
 
+    await notify_record_event(db, RecordEvent.updated, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/datasets.py]:
```suggestion
    await notify_record_event_v1(db, RecordEvent.updated, record)
```

File: argilla-server/src/argilla_server/contexts/datasets.py [hunk 12]
```diff
@@ -847,6 +850,8 @@ async def delete_record(db: AsyncSession, search_engine: "SearchEngine", record:
 
     await db.commit()
 
+    await notify_record_event(db, RecordEvent.deleted, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/datasets.py]:
```suggestion
    await notify_record_event_v1(db, RecordEvent.deleted, record)
```

File: argilla-server/src/argilla_server/contexts/datasets.py
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/datasets.py:1]: Btw there is a `delete_records` function here used by `/datasets/{dataset_id}/records` API endpoint. Do you want to cover that one too?

File: argilla-server/src/argilla_server/contexts/distribution.py [hunk 13]
```diff
@@ -21,6 +21,8 @@
 from sqlalchemy.orm import selectinload
 from sqlalchemy.ext.asyncio import AsyncSession
 
+from argilla_server.api.webhooks.v1.enums import RecordEvent
+from argilla_server.api.webhooks.v1.records import notify_record_event
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/distribution.py]:
```suggestion
from argilla_server.api.webhooks.v1.records import notify_record_event as notify_record_event_v1
```

File: argilla-server/src/argilla_server/contexts/distribution.py [hunk 14]
```diff
@@ -51,6 +53,11 @@ async def update_record_status(search_engine: SearchEngine, record_id: UUID) ->
 
         await db.commit()
 
+        await notify_record_event(db, RecordEvent.updated, record)
+
+        if record.is_completed():
+            await notify_record_event(db, RecordEvent.completed, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/distribution.py]:
```suggestion
            await notify_record_event_v1(db, RecordEvent.completed, record)
```

Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/distribution.py]: Are you sure we want to have a `record.updated` event and a `record.completed` event with the same info? I believe we should not notify the `record.updated` if we are gonna notify the `record completed` right?

But not 100% sure.

File: argilla-server/src/argilla_server/models/database.py [hunk 15]
```diff
@@ -200,6 +200,7 @@ def __repr__(self) -> str:
 
 class Record(DatabaseModel):
     __tablename__ = "records"
+    __table_args__ = (UniqueConstraint("external_id", "dataset_id", name="record_external_id_dataset_id_uq"),)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/models/database.py]: What is this?

File: argilla-server/src/argilla_server/contexts/distribution.py [hunk 16]
```diff
@@ -51,6 +53,11 @@ async def update_record_status(search_engine: SearchEngine, record_id: UUID) ->
 
         await db.commit()
 
+        await notify_record_event(db, RecordEvent.updated, record)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/contexts/distribution.py]:
```suggestion
        await notify_record_event_v1(db, RecordEvent.updated, record)
```

**Review by jfcalvo — CHANGES_REQUESTED** [review]

File: argilla-server/src/argilla_server/models/database.py (hunk 15 above)
Comment (frascuchon) [reply argilla-server/src/argilla_server/models/database.py]: I moved it from the previous line, but I realize that it is defined in the same place for other models. Rolling back this code formatting

File: argilla-server/src/argilla_server/contexts/distribution.py (hunk 14 above)
Comment (frascuchon) [reply argilla-server/src/argilla_server/contexts/distribution.py]: Let's try this way. We align the "standard" events and add the completed ones. Maybe we can rewrite this later.

File: argilla-server/src/argilla_server/api/webhooks/v1/records.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+
+async def notify_record_event(db: AsyncSession, record_event: RecordEvent, record: Record) -> List[Job]:
+    if record_event == RecordEvent.deleted:
+        return await _notify_record_deleted_event(db, record)
+
+    (
+        await db.execute(
+            select(Dataset)
+            .filter_by(id=record.dataset_id)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/webhooks/v1/records.py]: I remember some of us saying something like "let's start using `where` if possible" sometime ago:

```suggestion
            .where(Dataset.id == record.dataset_id)
```

**Review by jfcalvo — APPROVED** [review]

**Pull request merged**